autonat/README.md

```diff
@@ -0,0 +1,111 @@
+# NAT Discovery
+> How we dected if we're behind a NAT.
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Recommendation | Active | r0, 2021-08-26  |
+
+
+Authors: [@marten-seemann]
+
+Interest Group: [@mxinden], [@vyzo], [@raulk], [@stebalien], [@willscott]
+
+[@marten-seemann]: https://github.com/marten-seemann
+[@mxinden]: https://github.com/mxinden/
+[@vyzo]: https://github.com/vyzo
+[@raulk]: https://github.com/raulk
+[@stebalien]: https://github.com/stebalien
+[@willscott]: https://github.com/willscott
+
+See the [lifecycle document][lifecycle-spec] for context about maturity level
+and spec status.
+
+[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
+
+## Table of Contents
+
+TODO: fill in
+
+## Overview
+
+A priori, a node cannot know if it is behind a NAT or if it is publicly reachable.
+Knowing its NAT status is essential for the node to be well-behaved in the
+network: A node that's behind a NAT doesn't need to advertise its (undiable)
+addresses to the rest of the network, preventing superfluous dials from other
+peers. Furthermore, it might actively seek to improve its connectivity by
+finding a relay server, which would allow other peers to establish a relayed
+connection.
+
+To determine if it is located behind a NAT, nodes use the `autonat` protocol.
+Using this protocol, the node requests other peers to dial its presumed public
+addresses. If a couple of these dial attempts succeed, the node can be reasonably
+certain that it is not located behind a NAT. Likewise, if a couple of these dial
+attempts fail, this is a strong indicator that a NAT is blocking incoming
+connections.
+
+## AutoNAT Protocol
+
+The AutoNAT Protocol uses the Protocol ID `/libp2p/autonat/1.0.0`. The node
+wishing to determine its NAT status opens a stream using this protocol ID, and
+then sends a `Dial` message. The `Dial` message contains a list of multiaddresses.
+Upon receiving this message, the peer starts to dial these addresses. It MAY
+dial all of them in parallel. The peer MAY use a different IP and peer ID than
+it uses for its regular libp2p connection to perform these dial backs.
```

Comment:
If I understand correctly the fact that a server might use a different peer ID is due to an implementation detail in the Golang implementation. Given that AutoNAT is already widely deployed we won't be able to change that in this iteration. Still, would you mind documenting this, to ease discussions on this in future versions?

Reply:
I'm not sure we'd even want to change this in a future iteration, but I'm ok with documenting the reasoning here. Can you suggest some text?

Reply:
> I'm not sure we'd even want to change this in a future iteration

I might be missing something here. What is the benefit of this amendment other than not having to fix the shortcomings within a single implementation?

Reply:
I disagree that this is a shortcoming to begin with.
All we're interested in is if an address is dialable - who dials it (and from what address) is completely irrelevant.